### ISSUE TITLE
Panic on a specific flow of ops

### ISSUE THREAD

**Issue by leelavg** [helix-editor/helix#3722]
### Summary

- Panic in insert mode while move to bottom pane in tmux session
- I believe this may not be hit quite often, however will leave it to you

### Reproduction Steps

 

I tried this:

1. `RUST_BACKTRACE=full hx` in one of tmux panes (`hx` - zoomed in top pane and empty bottom pane)
2. Open a file in pwd via picker and another via `vs`
3. In insert mode edit the contents in right window and move to bottom tmux pane (mapped to ctrl-j)

I expected this to happen:
1. Move to tmux bottom pane and hx in top pane still running

Instead, this happened:
1. tmux top pane panicked
```
$ RUST_BACKTRACE=full hx
thread 'main' panicked at 'index out of bounds: the len is 3480 but the index is 5162', helix-tui/src/buffer.rs:495:33
stack backtrace:
   0:     0x564f92e0e6a4 - <unknown>
   1:     0x564f9262e84c - <unknown>
   2:     0x564f92e083e5 - <unknown>
   3:     0x564f92e1020a - <unknown>
   4:     0x564f92e0ff05 - <unknown>
   5:     0x564f92e10912 - <unknown>
   6:     0x564f92e10734 - <unknown>
   7:     0x564f92e0ebd4 - <unknown>
   8:     0x564f92e1049d - <unknown>
   9:     0x564f925a5ca3 - <unknown>
  10:     0x564f925a5be2 - <unknown>
  11:     0x564f92b916c8 - <unknown>
  12:     0x564f929e93ae - <unknown>
  13:     0x564f92b0b092 - <unknown>
  14:     0x564f929c9535 - <unknown>
  15:     0x564f929bded4 - <unknown>
  16:     0x564f929dd148 - <unknown>
  17:     0x564f92c8dfe8 - <unknown>
  18:     0x564f92caa0bc - <unknown>
  19:     0x564f92c99f45 - <unknown>
  20:     0x564f92c75828 - <unknown>
  21:     0x564f92c5e398 - <unknown>
  22:     0x564f92cb8b20 - <unknown>
  23:     0x564f92c78586 - <unknown>
  24:     0x564f92c6f9a3 - <unknown>
  25:     0x564f92cb0151 - <unknown>
  26:     0x564f92e030d2 - <unknown>
  27:     0x564f92c78692 - <unknown>
  28:     0x7fc97dc29550 - __libc_start_call_main
  29:     0x7fc97dc29609 - __libc_start_main_impl
  30:     0x564f925dc23e - <unknown>
  31:                0x0 - <unknown>
```

### Helix log

- I removed the file before reproducing and can only see below
- In verbose mode it logs a lot however there are only INFO logs from LSP

<details><summary>~/.cache/helix/helix.log</summary>

```
$ cat ~/.cache/helix/helix.log 
2022-09-06T09:38:14.708 helix_lsp::transport [ERROR] err: <- StreamClosed
```

</details>

### Platform

Linux

### Terminal Emulator

foot version: 1.13.0 -pgo +ime +graphemes -assertions

### Helix Version

helix 22.08.1 (66276ce6)

**Comment by leelavg** [comment]
asciinema: [hx-bug.zip](https://github.com/helix-editor/helix/files/9493215/hx-bug.zip)

**Comment by vvzen** [comment]
I get something along these lines (panicking after an index out of bounds) a few times per week, sadly it's very hard to make a clear MRE since it always happens when I'm deep into coding and I press a specific combination of keys, I'll try my best the next time it happens.
When it happens and you haven't saved your current buffer, it can be a bit annoying.

I'm on `helix 22.12 (96ff64a8)`.

This is my full backtrace:
```bash
$ hx src/core/dev.rs
thread 'main' panicked at 'index out of bounds: the len is 6516 but the index is 8204', helix-tui/src/buffer.rs:513:33
stack backtrace:
   0:        0x1055f69cc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h71d20ef014883dc3
   1:        0x104e0d9a4 - core::fmt::write::hd114a9636e34e383
   2:        0x1055e06c4 - std::io::Write::write_fmt::h05898edc92938c00
   3:        0x1055e6cf8 - std::panicking::default_hook::{{closure}}::hbd40827d346a3df4
   4:        0x1055e6748 - std::panicking::default_hook::hf2e0d2c157a06c84
   5:        0x1055e7594 - std::panicking::rust_panic_with_hook::h33fd2dae40c879df
   6:        0x1055f706c - std::panicking::begin_panic_handler::{{closure}}::h81ef06081ce645db
   7:        0x1055f6afc - std::sys_common::backtrace::__rust_end_short_backtrace::hc514eeb5ddb8eaf3
   8:        0x1055e711c - _rust_begin_unwind
   9:        0x10568a224 - core::panicking::panic_fmt::h520fb0ef56e6cee7
  10:        0x10568a194 - core::panicking::panic_bounds_check::hde40779dd2c2d133
  11:        0x105360f3c - helix_tui::buffer::Buffer::clear_with::h32c92af2ff926d15
  12:        0x10519cb48 - <helix_term::ui::popup::Popup<T> as helix_term::compositor::Component>::render::hba322a4f95a944ee
  13:        0x1052c9b7c - <helix_term::ui::completion::Completion as helix_term::compositor::Component>::render::h0d43d930eb2aac6b
  14:        0x1051c64c8 - <helix_term::ui::editor::EditorView as helix_term::compositor::Component>::render::h6dadc24a8a7b900a
  15:        0x105486b04 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hbdb5b350f914e87a
  16:        0x10546b130 - helix_term::application::Application::run::{{closure}}::ha1a6a3d2e9110b6f
  17:        0x105485da8 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::ha293dcda9e9e5786
  18:        0x1054b8c50 - tokio::runtime::park::CachedParkThread::block_on::h51e6522aa4a46641
  19:        0x1054d8b64 - tokio::runtime::runtime::Runtime::block_on::h9b47b4af438e10d8
  20:        0x105487434 - hx::main::hb1d93c2fc2e8afc9
  21:        0x1054bfac8 - std::sys_common::backtrace::__rust_begin_short_backtrace::hd520b3637b9dd87a
  22:        0x1054cf9d8 - std::rt::lang_start::{{closure}}::h06c2ba16be546c7c
  23:        0x1055e5384 - std::rt::lang_start_internal::h3f58ab0124dae34f
  24:        0x105487544 - _main
```

Thanks!

**Comment by Plasma-Vortex** [comment]
> it always happens when I'm deep into coding and I press a specific combination of keys

Does this bug occur right when you resize your terminal window?

**Comment by vvzen** [comment]
> Does this bug occur right when you resize your terminal window? And are you also using tmux?
I rarely resize my terminal window since I use Amethyst (a window manager that I use to force my window to stay fullscreen) so I don't think it's necessarily tied to that, but yes I do use tmux!

Do you think there are any side effects in switching to a safer index accessing ?

**Comment by pascalkuthe** [comment]
Sounds like this is likely fixed by #5529

**Comment by leelavg** [comment]
Yep, mostly the popup not having enough space is a good generalization for what I faced.

Earlier attached asciinema also shows a similar symptom.

**Comment by pascalkuthe** [comment]
This was likely fixed by  #6109, closing
